pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,140 @@
+
+	KubeletVersionUnknownReason           = "KubeletVersionUnknown"
+	KubeletVersionSyncedReason            = "KubeletVersionSynced"
+	KubeletVersionSkewLessThanLimitReason = "KubeletVersionSkewLessThanLimit"
+	KubeletVersionSkewAtLimitReason       = "KubeletVersionSkewAtLimit"
+	KubeletVersionUnsupportedReason       = "KubeletVersionUnsupported"
+)
+
+// KubeletVersionSkewController sets Upgradeable=False if the kubelet
+// version on a node is two minor versions older than the API server
+// in order to prevent upgrading to an unsupported API server version.
+type KubeletVersionSkewController interface {
+	factory.Controller
+}
+
+func NewKubeletVersionSkewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:   operatorClient,
+		nodeLister:       kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion: semver.MustParse(status.VersionForOperandFromEnv()),
+	}
+	// check if major version updated and no one updated this code
+	if c.apiServerVersion.Minor < 2 {
+		// minor version was 0 or 1, which most likely means the major version has been updated
+		panic(fmt.Errorf("unable to determine kubelet version skew limit"))
+	}
+	c.Controller = factory.New().
+		WithSync(c.sync).
+		WithInformers(kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Informer()).
+		ToController("KubeletVersionSkewController", recorder.WithComponentSuffix("kubelet-version-skew-controller"))
+	return c
+}
+
+type kubeletVersionSkewController struct {
+	factory.Controller
+	operatorClient   v1helpers.OperatorClient
+	nodeLister       corev1listers.NodeLister
+	apiServerVersion semver.Version
+}
+
+func (c *kubeletVersionSkewController) sync(_ context.Context, _ factory.SyncContext) error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(operatorSpec.ManagementState) {
+		return nil
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errorMsgs []string
+	var skewedToTheLimitMsgs []string
+	var skewedButOKMsgs []string
+	var syncedMsgs []string
+	var unsupportedMsgs []string
+
+	// for each node, check kubelet version
+	for _, node := range nodes {
+		kubeletVersion, err := nodeKubeletVersion(node)
+		skew := int(kubeletVersion.Minor - c.apiServerVersion.Minor)
+		switch {
+		case err != nil:
+			errorMsgs = append(errorMsgs, fmt.Sprintf("unable to determine kubelet version on node %s: %s", node.Name, err))
+		case skew <= -2: // behind too far
+			skewedToTheLimitMsgs = append(skewedToTheLimitMsgs, fmt.Sprintf("Kubelet version (%s) on node %s is behind the expected API server version (%s) by at least the maximum allowable skew.", kubeletVersion, node.Name, c.apiServerVersion))
```

Comment:
kubelet.minor = `1` and apiserver.minor = `4` -> skew = 1 - 4 = -3.  `-3 <= -2 = true` so is `-2 <= -2` . Are we putting skew of -3 and -2 in the same category as `skewedToTheLimitMsgs`. I thought a skew of `2` is acceptable.

I thought it would be 
```
switch {
case skew > 0:
  // kubelet version is newer, rollack
case skew == 0:
  // synced
case skew >= -2:
  // skew within limit of 2
default:
  // too far
}
```

May be add some comments to clarify this?

Reply:
> Are we putting skew of -3 and -2 in the same category as `skewedToTheLimitMsgs`. I thought a skew of `2` is acceptable.

Skew of `-2` is supported, but this is about preventing you from upgrading to the the **next** minor version of API server, in which case you would end up in an unsupported state, thus the `-2` is included in the same case as the `-3`.

Reply:
Added clarifications to the comments.

Reply:
Currently this case includes both "still supported, but `Upgradeable=False` because bumping would leave you unsupported" and "you are unsupported right now".  I'm trying to wrap my head around what I think I want instead...

Reply:
I think we need to directly address the expected Kube-API version of the next OpenShift minor, unless we can just assume that each OpenShift minor bump will come with a matching Kube-API bump.  Even if we do assume that, writing it down in a line of Go would make things more clear:
 
 ```go
expectedSkewVsNextOpenShiftMinor := skew - 1 // assume that every OpenShift minor bump includes a Kube API minor bump
```

Then:

* `skew == 0`: synced.  Fill `syncedMsgs`.
* `skew < c.minUpgradeableSkew`: so far back it's currently unsupported.  Fill `unsupportedMsgs`.
* `expectedSkewVsNextOpenShiftMinor <= c.minUpgradeableSkewNextMinor`: upgrading to next minor version of API server would result in an unsupported config.  Fill `skewedToTheLimitMsgs`.
* `skew < 0`: behind, but upgrading to next minor version of API server is supported.  Fill `skewedButOKMsgs`.
* `default`: kubelet version newer than API server version.  Possibly in the middle of a rollback.  Fill `unsupportedMsgs`.

Working some examples, fiddling with the even/odd `minUpgradeableSkew` settings vs. your current `minUpgradeableKubeletSkewForOCPVersion` implementation:

* OCP 4.9, `minUpgradeableSkew = -1; minUpgradeableSkewNextMinor = -2`.  API 1.22 (expected next API 1.23).  Kubelet 1.22, `skew = 0; expectedSkewVsNextOpenShiftMinor = -1`.  `syncedMsgs`.
* OCP 4.9, `minUpgradeableSkew = -1; minUpgradeableSkewNextMinor = -2`.  API 1.22 (expected next API 1.23).  Kubelet 1.21, `skew = -1; expectedSkewVsNextOpenShiftMinor = -2`.  `skewedButOKMsgs`.
* OCP 4.9, `minUpgradeableSkew = -1; minUpgradeableSkewNextMinor = -2`.  API 1.22 (expected next API 1.23).  Kubelet 1.20, `skew = -2; expectedSkewVsNextOpenShiftMinor = -3`.  `unsupportedMsgs`.
* OCP 4.8, `minUpgradeableSkew = -2; minUpgradeableSkewNextMinor = -1`.  API 1.21 (expected next API 1.22).  Kubelet 1.21, `skew = 0; expectedSkewVsNextOpenShiftMinor = -1`.  `syncedMsgs`.
* OCP 4.8, `minUpgradeableSkew = -2; minUpgradeableSkewNextMinor = -1`.  API 1.21 (expected next API 1.22).  Kubelet 1.20, `skew = -1; expectedSkewVsNextOpenShiftMinor = -2`.  `skewedToTheLimitMsgs`.
* OCP 4.8, `minUpgradeableSkew = -2; minUpgradeableSkewNextMinor = -1`.  API 1.21 (expected next API 1.22).  Kubelet 1.19, `skew = -2; expectedSkewVsNextOpenShiftMinor = -3`.  `skewedToTheLimitMsgs`.
* OCP 4.8, `minUpgradeableSkew = -2; minUpgradeableSkewNextMinor = -1`.  API 1.21 (expected next API 1.22).  Kubelet 1.18, `skew = -3; expectedSkewVsNextOpenShiftMinor = -4`.  `unsupported`.

Reply:
> Currently this case includes both "still supported, but `Upgradeable=False` because bumping would leave you unsupported" and "you are unsupported right now". 

It's not clear which version of the switch you are referring to as it has changed a bit, so pasting the current one here w/ comments.

```go
skew := int(kubeletVersion.Minor - c.apiServerVersion.Minor)
switch {
case skew == 0:
	// Minor versions synced.
	syncedMsgs = append(syncedMsgs, ...)
case skew < c.minUpgradeableSkew:
	// Upgrading to next minor version of OCP would result in an unsupported OCP state.
	// Note that we might already be in an unsupported OCP state (kubelet too far
	// behind), but the end result is the same, we don't want you to upgrade to the
	// next minor version of OCP.
	skewedToTheLimitMsgs = append(skewedToTheLimitMsgs, ...)
case skew < 0:
	// Behind, but upgrading to next minor version of API server is supported.
	// This can only occur when OCP minor version is odd.
	skewedButOKMsgs = append(skewedButOKMsgs, ...)
default:
	// Kubelet version newer than api server version for the current OCP version.
	// Possibly in the middle of a rollback. Let's just stay out of the way.
	unsupportedMsgs = append(unsupportedMsgs, ...)
}
```

I don't see the need to split the `skew < c.minUpgradeableSkew` case into _"unsupported already"_ and _"unsupported if updated"_ for the scope of this particular controller as the result is still the same: `Upgradeable=False`.

Reply:
> ...for the scope of this particular controller as the result is still the same: `Upgradeable=False`.

In both cases, `Upgradeable` should be `False`.  But in the will-become-unsupported case, the messaging should be "not a big deal, but you'll want to fix this before you can bump minors", while the currently-unsupported case should be "this is very bad right now.  [Nasal demons)][1]".  And ideally there would be an alert going off... but we can come back to look at the alert if we actually see this in condition Telemetry (another reason to have different reasons for the "nodes too far in front" and "nodes too far behind" cases, and possibly also a reason for "both at once, since they are both equally serious").

[1]: https://en.wikipedia.org/wiki/Undefined_behavior

Reply:
For the immediate future, we can assume every openshift comes with a new kube and that we never skip a kube.  If that assumption breaks in the future, we can backport a specific change to directly handle it.

Reply:
See https://github.com/openshift/cluster-kube-apiserver-operator/pull/1199#discussion_r693237388 for a follow-up.